discovery/src/main/java/org/ethereum/beacon/discovery/enr/EnrSchemeInterpreter.java

```diff
@@ -0,0 +1,23 @@
+package org.ethereum.beacon.discovery.enr;
+
+import org.web3j.rlp.RlpString;
+import tech.pegasys.artemis.util.bytes.Bytes32;
+
+public interface EnrSchemeInterpreter {
+  /** Returns supported scheme */
+  EnrScheme getScheme();
+
+  /** Verifies that `nodeRecord` is of scheme implementation */
+  default void verify(NodeRecord nodeRecord) {
+    if (!nodeRecord.getIdentityScheme().equals(getScheme())) {
+      throw new RuntimeException("Interpreter and node record schemes do not match!");
+    }
+  }
+
+  /** Delivers nodeId according to identity scheme scheme */
+  Bytes32 getNodeId(NodeRecord nodeRecord);
+
+  Object decode(String key, RlpString rlpString);
```

Comment:
Identity scheme is not responsible for encode/decode. Why would we want to add this responsibility here?

If you want to make an abstraction for ENR codec you may do this with a separated interface. The only implementation that makes sense so far is an RLP codec.

Discovery might be initialized with identity scheme factory that in its turn is initialized with a list of supported schemes. When `NodeRecord` gets parsed it's passed on identity schema to get verified; also, identity scheme is used to derive node id. It could be done in a factory method as you already do it in the following code.